src/wasm/wasm-emscripten.cpp

```diff
@@ -305,32 +297,114 @@ void AsmConstWalker::process() {
   addImports();
 }
 
-void AsmConstWalker::createAsmConst(uint64_t id, std::string code) {
-  AsmConst asmConst;
-  asmConst.id = id;
-  asmConst.code = code;
-  asmConsts.push_back(asmConst);
-}
-
 void AsmConstWalker::addImports() {
   for (auto& import : queuedImports) {
     wasm.addFunction(import.release());
   }
 }
 
-static AsmConstWalker findEmAsmConstsAndReturnWalker(Module& wasm,
-                                                     bool minimizeWasmChanges) {
-  AsmConstWalker walker(wasm, minimizeWasmChanges);
-  walker.process();
-  return walker;
+struct SegmentRemover : WalkerPass<PostWalker<SegmentRemover>> {
+  SegmentRemover(Index segment) : segment(segment) {}
+
+  bool isFunctionParallel() override { return true; }
+
+  Pass* create() override { return new SegmentRemover(segment); }
+
+  void visitMemoryInit(MemoryInit* curr) {
+    if (segment == curr->segment) {
+      Builder builder(*getModule());
+      replaceCurrent(builder.blockify(builder.makeDrop(curr->dest),
+                                      builder.makeDrop(curr->offset),
+                                      builder.makeDrop(curr->size)));
+    }
+  }
+
+  void visitDataDrop(DataDrop* curr) {
+    if (segment == curr->segment) {
+      Builder builder(*getModule());
+      replaceCurrent(builder.makeNop());
+    }
+  }
+
+  Index segment;
+};
+
+static void removeSegment(Module& wasm, Index segment) {
+  PassRunner runner(&wasm);
+  SegmentRemover(segment).run(&runner, &wasm);
+  // Resize the segment to zero.  In theory we should completely remove it
+  // but that would mean re-numbering the segments that follow which is
+  // non-trivial.
+  wasm.memory.segments[segment].data.resize(0);
+}
+
+static Address getExportedAddress(Module& wasm, Export* export_) {
+  Global* g = wasm.getGlobal(export_->value);
+  auto* addrConst = g->init->dynCast<Const>();
+  return addrConst->value.getInteger();
+}
+
+static std::vector<AsmConst> findEmAsmConsts(Module& wasm,
+                                             bool minimizeWasmChanges) {
+  Export* start = wasm.getExportOrNull("__start_em_asm");
+  Export* end = wasm.getExportOrNull("__stop_em_asm");
+
+  // Older versions of emscripten don't export these symbols.  Instead
+  // we run AsmConstWalker in an attempt to derive the string addresses
+  // from the code.
+  if (!start || !end) {
+    AsmConstWalker walker(wasm, minimizeWasmChanges);
+    walker.process();
+    return walker.asmConsts;
+  }
```

Comment:
Are planning to remove this fallback code in the future? What is stopping us from removing it now?

Reply:
Yes.   Mostly its a transition thing.    I can't land the emscripten-side change first because it actually changes the way that arg0 gets passed the em_asm functions.   llvm ends up storing that address of the string in memory and then loading it back (at least in -O0) which is not currently supported by the AsmConstWalker.   So if I land the emscripten side first it would break this pass.